language-server/modules/langserver-commons/src/main/java/org/ballerinalang/langserver/commons/toml/common/completion/TomlCompletionUtil.java

```diff
@@ -0,0 +1,233 @@
+     *
+     * @param topLevelNode The toplevel node to which the completion items belong to.
+     * @param snippets     Set of completion items to be filtered.
+     * @return {@link Map<String,CompletionItem> } The set of filtered completion items.
+     */
+    public static Map<String, CompletionItem> getFilteredCompletions(Either<TableNode,
+            TableArrayNode> topLevelNode, Map<TomlNode, Map<String, CompletionItem>> snippets) {
+
+        String tableKey;
+        NodeList<KeyValueNode> fields;
+        DocumentNode documentNode;
+        if (topLevelNode.isLeft()) {
+            tableKey = TomlSyntaxTreeUtil.toQualifiedName(topLevelNode.getLeft().identifier().value());
+            fields = topLevelNode.getLeft().fields();
+            documentNode = (DocumentNode) topLevelNode.getLeft().parent();
+        } else {
+            tableKey = TomlSyntaxTreeUtil.toQualifiedName(topLevelNode.getRight().identifier().value());
+            fields = topLevelNode.getRight().fields();
+            documentNode = (DocumentNode) topLevelNode.getRight().parent();
+        }
+
+        Map<String, CompletionItem> completions =
+                new HashMap<>(findCompletionItemsFromQualifiedKey(snippets, tableKey));
+
+        /*   Removes the following.
+            eg: [build-options]
+                cloud="docker"
+            The completion item that corresponds to the above key-value pair
+            will be removed as it is already defined.
+         */
+        for (KeyValueNode field : fields) {
+            String key = TomlSyntaxTreeUtil.toQualifiedName(field.identifier().value());
+            completions.remove(key);
+        }
+
+        /*
+            Add all possible top-level (Table & TableArray) keys to the completion-item list.
+            And remove already existing keys from the completion item list.
+            eg: [build-options],[package]
+         */
+        addTopLevelNodeCompletions(snippets.keySet(), completions);
+        return removeExistingTableKeys(completions, documentNode);
+    }
+
+    /**
+     * Returns the set of  completion items for the given set of toml nodes.
+     *
+     * @param topLevelNodes The set of top-level toml nodes.
+     * @param completions   The set of completion items.
+     * @return
+     */
+    public static Map<String, CompletionItem> addTopLevelNodeCompletions(Set<TomlNode> topLevelNodes,
+                                                                            Map<String, CompletionItem> completions) {
+        for (TomlNode topLevelNode : topLevelNodes) {
+            String key = topLevelNode.getKey();
+            if (completions.containsKey(key)) {
+                continue;
+            }
+            CompletionItem item;
+            if (topLevelNode.type() == TomlNodeType.TABLE || topLevelNode.type() == TomlNodeType.TABLE_ARRAY) {
+                item = createTopLevelCompletionItem(topLevelNode);
+            } else {
+                continue;
+            }
```

Comment:
Shall we merge the two `continue` cases?

Reply:
Done